### ISSUE TITLE
Problems with training my unconditional model 

### ISSUE THREAD

**Issue by benfig1127** [mikeizbicki/cmc-csci181#23]
I am having problems with reproducing the results of the "bad accuracy" for the unconditional model. Whenever I try and train my model with the ```--conditional_model``` flag it appears there is something wrong with the way my model is looking at the criteria, but I cannot figure out what is wrong in the training loop. Here is the output of my tensorboard after 7000 steps 
![Screenshot from 2020-04-14 19-23-46](https://user-images.githubusercontent.com/61296503/79292276-8c6e2400-7e85-11ea-9628-b8bd1db47d8f.png)
![Screenshot from 2020-04-14 19-23-50](https://user-images.githubusercontent.com/61296503/79292284-8e37e780-7e85-11ea-9ac3-b8202063ce09.png)
As you can see these are converging far to fast to extremly high accuracy values without the ```loss_nextchars``` improving enough. 
Here are the exact command line arguments I am using:
```
python3 names.py --train --model=lstm --conditional_model --gradient_clipping --optimizer=adam --samples=50000 --learning_rate=1e-1 --batch_size=10 --model_name=model_hw7_6_1 
```
The training procedure works fine if I leave the ```--conditional_model``` flag off, so I believe the problem stems in the following code somewhere, but I cannot figure out where it is. 
My code: 
```
# training
if args.train:
    print('Training! Please wait')
    # prepare model for training
    criterion = nn.CrossEntropyLoss()
    if args.optimizer == 'sgd':
        optimizer = torch.optim.SGD(
                model.parameters(),
                lr=args.learning_rate,
                momentum=args.momentum,
                weight_decay=args.weight_decay
                )
    if args.optimizer == 'adam':
        optimizer = torch.optim.Adam(
            model.parameters(),
            lr=args.learning_rate,
            weight_decay=args.weight_decay
             )
    if args.optimizer =='adadelta':
        optimizer = torch.optim.Adadelta(
            model.parameters(),
            lr=args.learning_rate,
            weight_decay=args.weight_decay
            )
    model.train()
```
```
    # training loop
    start_time = time.time()
    for step in range(1, args.samples + 1):
        # get random training example
        category_list=[]
        line_list=[]
        
        for i in range(0,args.batch_size):
            category = random.choice(all_categories)
            line = random.choice(category_lines[category])
            category_list.append(all_categories.index(category))
            line_list.append(line)
            
        category_tensor = torch.tensor(category_list, dtype=torch.long)
        line_tensor = str_to_tensor(line_list)


        if args.conditional_model:    
            category_onehot=torch.nn.functional.one_hot(category_tensor, len(all_categories)).float()
            category_onehot=torch.unsqueeze(category_onehot,0)
            category_onehot=torch.cat(line_tensor.shape[0]*[category_onehot],dim=0)
            input_tensor=torch.cat([line_tensor,category_onehot],dim=2)
        else:
            input_tensor=line_tensor
            
        # perform training step
        output_class, output_nextchars = model(input_tensor)
        loss_class = criterion(output_class, category_tensor)
        loss_nextchars_perchar= torch.zeros(output_nextchars.shape[0])
        for i in range(output_nextchars.shape[0]-1):
            _, nextchar_i=line_tensor[i+1,:].topk(1)
            nextchar_i=nextchar_i.view(-1)
            loss_nextchars_perchar[i]=criterion(output_nextchars[i,:],nextchar_i)
        loss_nextchars=torch.mean(loss_nextchars_perchar)
        
        loss=loss_class +loss_nextchars
        loss.backward()
        grad_norm = sum([ torch.norm(p.grad)**2 for p in model.parameters() if p.grad is not None])**(1/2)
        if args.gradient_clipping:
            torch.nn.utils.clip_grad_norm_(model.parameters(),1.0)
        optimizer.step()
        
        
        
        
        # get category from output
        top_n, top_i = output_class.topk(1)
        guess_i = top_i[-1].item()
        category_i = category_tensor[-1]
        guess = all_categories[guess_i]
        category = all_categories[category_i]
        accuracies = torch.where(
                top_i[:,0]==category_tensor,
                torch.ones([args.batch_size]),
                torch.zeros([args.batch_size])
                )
        accuracy = torch.mean(accuracies).item()
        
        # tensorboard
        writer.add_scalar('train/loss_class', loss_class.item(), step)
        writer.add_scalar('train/loss_nextchars', loss_nextchars.item(), step)
        writer.add_scalar('train/loss', loss.item(), step)
        writer.add_scalar('train/accuracy', accuracy, step)
        writer.add_scalar('train/grad_norm', grad_norm.item(), step)

        # print status update
        if step % 100 == 0:
            correct = '✓' if guess == category else '✗ (%s)' % category
            print('%d %d%% (%.2f sec) %.4f %s / %s %s' % (
                step,
                step / args.samples * 100,
                time.time()-start_time,
                loss,
                line,
                guess,
                correct
                ))

        # save model
        if step%args.save_every == 0 or step==args.samples:
            print('saving model checkpoint')
            torch.save({
                    'step':step,
                    'model_state_dict': model.state_dict(),
                    'optimizer_state_dict': optimizer.state_dict(),
                    'loss':loss
                }, os.path.join(log_dir,'model'))

```

**Comment by tadizov** [comment]
I think it might be because you using the sum of class loss and nextchars loss as your loss function while training the model. You already know your class for the conditional model, so thats probably why you get such high accuracy from the very beginning ..but I am not 100% sure., try setting loss function just to loss of nextchar when passing --conditional_model.

**Comment by benfig1127** [comment]
@tadizov Great point! I am not sure how I overlooked that, let me change it and see what happens :).

**Comment by benfig1127** [comment]
@tadizov It seems to have fixed it, thanks so much!
